src/python/pants/pantsd/pailgun_server.py

```diff
@@ -178,19 +195,74 @@ def handle_request(self):
     with self.lifecycle_lock():
       self._handle_request_noblock()
 
+  def _should_poll_forever(self, timeout):
+    return timeout < 0
+
+  def _should_keep_polling(self, timeout, time_polled):
+    return self._should_poll_forever(timeout) or time_polled < timeout
+
+  @contextmanager
+  def ensure_request_is_exclusive(self, environment, request):
+    """
+    Ensure that this is the only pants running.
+
+    We currently don't allow parallel pants runs, so this function blocks a request thread until
+    there are no more requests being handled.
+    """
+    # TODO add `did_poll` to pantsd metrics
+
+    timeout = float(environment['PANTSD_REQUEST_TIMEOUT_LIMIT'])
+
+    @contextmanager
+    def yield_and_release(time_waited):
+      try:
+        self.logger.debug("request lock acquired {}.".format("on the first try" if time_waited == 0 else "in {} seconds".format(time_waited)))
+        self.free_to_handle_request_lock.clear()
```

Comment:
I don't know `threading.Event`, but from reading its documentation it seems like nothing guarantees that only one thread can clear an `Event` at a time. I suspect you may either need to either guard the `clear`ing in a lock, or use a `threading.Condition` (or possibly `threading.Semaphore`, or just a `threading.Lock` with `blocking=False`) to ensure that only one thread can be woken up at a time.

It would be good to change the test to coordinate 3 or 10 threads, rather than 2, to make sure this works for n>2.

Reply:
Changed for a temporary implementation of `Lock` with `Conditional` and a boolean. This should be gone after Py3 is the only thing we care about, but for now it will do.